### ISSUE TITLE
cross compile for i686 (Qnap NAS failed)

### ISSUE THREAD

**Issue by nephaste** [weidai11/cryptopp#822]
I am trying to cross compile from my ubuntu1404 i686 32bit (LXC) for a Qnap NAS (GCC 4.3.6 i686  32bit)... to include it for UrBackup 

I am stuck on this error

```
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -msse2 -c donna_sse.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c dsa.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c eax.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c ec2n.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c eccrypto.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c ecp.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c elgamal.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c emsa2.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c eprecomp.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c esign.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c files.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c filters.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c fips140.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c fipstest.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -c gcm.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -I/opt/QUrBackup/include -msse2 -c gcm_simd.cpp
In file included from gcm_simd.cpp:28:
/data/i686-QNAP-linux-gnu/bin/../lib/gcc/i686-QNAP-linux-gnu/4.3.6/include/tmmintrin.h:34:3: error: #error "SSSE3 instruction set not enabled"
gcm_simd.cpp:29:24: error: wmmintrin.h: No such file or directory
gcm_simd.cpp: In function 'long long int __vector__ CryptoPP::GCM_Reduce_CLMUL(long long int __vector__, long long int __vector__, long long int __vector__, const long long int __vector__&)':
gcm_simd.cpp:447: error: '_mm_clmulepi64_si128' was not declared in this scope
gcm_simd.cpp: In function 'long long int __vector__ CryptoPP::GCM_Multiply_CLMUL(const long long int __vector__&, const long long int __vector__&, const long long int __vector__&)':
gcm_simd.cpp:463: error: '_mm_clmulepi64_si128' was not declared in this scope
gcm_simd.cpp: In function 'void CryptoPP::GCM_SetKeyWithoutResync_CLMUL(const CryptoPP::byte*, CryptoPP::byte*, unsigned int)':
gcm_simd.cpp:474: error: '_mm_shuffle_epi8' was not declared in this scope
gcm_simd.cpp: In function 'size_t CryptoPP::GCM_AuthenticateBlocks_CLMUL(const CryptoPP::byte*, size_t, const CryptoPP::byte*, CryptoPP::byte*)':
gcm_simd.cpp:505: error: '_mm_shuffle_epi8' was not declared in this scope
gcm_simd.cpp:520: error: '_mm_clmulepi64_si128' was not declared in this scope
gcm_simd.cpp:528: error: '_mm_clmulepi64_si128' was not declared in this scope
gcm_simd.cpp: In function 'void CryptoPP::GCM_ReverseHashBufferIfNeeded_CLMUL(CryptoPP::byte*)':
gcm_simd.cpp:566: error: '_mm_shuffle_epi8' was not declared in this scope
make: *** [gcm_simd.o] Error 1
```

Have any idea on how to fix ?

**Comment by noloader** [comment]
Add `-DCRYPTOPP_DISABLE_CLMUL` to your command line. Or better, set it in `config.h` so that the library and programs that link to the library have the same definition.

**Comment by nephaste** [comment]
thanks it works... but new error come i had disable SSSE3 as well

but i ma stuck on SHAKE error ... tried -DCRYPTOPP_DISABLE_SHAKE and also -DCRYPTOPP_DISABLE_SHAKE128   ... has no effect

> 
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c seal.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c seed.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c serpent.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c sha.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c sha3.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c sha_simd.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c shacal2.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c shacal2_simd.cpp
i686-QNAP-linux-gnu-g++ -O2 -fPIC -DCRYPTOPP_DISABLE_CLMUL -DCRYPTOPP_DISABLE_SSSE3 -DCRYPTOPP_DISABLE_SHAKE -I/opt/QUrBackup/include -c shake.cpp
In file included from shake.cpp:21:
shake.h: In constructor 'CryptoPP::SHAKE128::SHAKE128(unsigned int)':
shake.h:128: error: class 'CryptoPP::SHAKE128' does not have any field named 'SHAKE_Final'
shake.h: In constructor 'CryptoPP::SHAKE256::SHAKE256(unsigned int)':
shake.h:158: error: class 'CryptoPP::SHAKE256' does not have any field named 'SHAKE_Final'
make: *** [shake.o] Error 1


any clue ?

**Comment by noloader** [comment]
You should only need `-DCRYPTOPP_DISABLE_SSSE3`. It disables SSSE3 and above. Also see [`GNUmakefile` : 365](https://github.com/weidai11/cryptopp/blob/master/GNUmakefile#L366) or so.

This fixes the SHAKE128 problem: [Commit fc3b16e3a39d](https://github.com/weidai11/cryptopp/commit/fc3b16e3a39d).

**Comment by nephaste** [comment]
thanks again...

Urbackup working seamlessly now  ;)
